### ISSUE TITLE
"ImportError: PYZ entry xxx failed to decompress" in Gunicorn's worker process

### ISSUE THREAD

**Issue by hyt-hz** [pyinstaller/pyinstaller#1070]
My Gunicorn application uses gevent wsgi server as its worker, and gunicorn forks multiple workers to better utilize multi-core CPUs. I bundled my application in one folder, but the bundled application failed randomly when starts up because the worker process failed immediately after forking.

<pre><code>
[2014-11-01 05:38:38 +0000] [872] [INFO] Booting worker with pid: 872
[2014-11-01 05:38:38 +0000] [870] [ERROR] Exception in worker process:
Traceback (most recent call last):
  File "/mnt/gss_hep/pyinstaller/build/gss/out00-PYZ.pyz/gunicorn.arbiter", line 507, in spawn_worker
  File "/mnt/gss_hep/pyinstaller/build/gss/out00-PYZ.pyz/gunicorn.workers.ggevent", line 193, in init_process
  File "/mnt/gss_hep/pyinstaller/build/gss/out00-PYZ.pyz/gunicorn.workers.base", line 114, in init_process
  File "/mnt/gss_hep/pyinstaller/build/gss/out00-PYZ.pyz/gunicorn.app.base", line 66, in wsgi
  File "<string>", line 42, in load
  File "/home/hyt/pyenv/lib/python2.7/site-packages/PyInstaller/loader/pyi_importers.py", line 270, in load_module
    exec(bytecode, module.__dict__)
  File "/mnt/gss_hep/pyinstaller/build/gss/out00-PYZ.pyz/gss.api.rest", line 9, in <module>
  File "/home/hyt/pyenv/lib/python2.7/site-packages/PyInstaller/loader/pyi_importers.py", line 270, in load_module
    exec(bytecode, module.__dict__)
  File "/mnt/gss_hep/pyinstaller/build/gss/out00-PYZ.pyz/flask", line 19, in <module>
  File "/home/hyt/pyenv/lib/python2.7/site-packages/PyInstaller/loader/pyi_importers.py", line 270, in load_module
    exec(bytecode, module.__dict__)
  File "/mnt/gss_hep/pyinstaller/build/gss/out00-PYZ.pyz/jinja2", line 33, in <module>
  File "/home/hyt/pyenv/lib/python2.7/site-packages/PyInstaller/loader/pyi_importers.py", line 270, in load_module
    exec(bytecode, module.__dict__)
  File "/mnt/gss_hep/pyinstaller/build/gss/out00-PYZ.pyz/jinja2.environment", line 21, in <module>
  File "/home/hyt/pyenv/lib/python2.7/site-packages/PyInstaller/loader/pyi_importers.py", line 202, in load_module
    is_pkg, bytecode = self._pyz_archive.extract(fullname)
  File "/home/hyt/pyenv/lib/python2.7/site-packages/PyInstaller/loader/pyi_archive.py", line 315, in extract
    raise ImportError("PYZ entry '%s' failed to decompress" % name)
ImportError: PYZ entry 'jinja2.parser' failed to decompress
</pre></code>

The failed PYZ entry is random, and if set worker process to 1 (only forks 1 worker process), everyting is fine, so I am guessing there is some kind of race condition concerning the opened ZlibArchive file, since all the worker processors share the same fd, and there might be chance they are seeking and reading this file at the same time, when trying to import module from it.

**Comment by hyt-hz** [comment]
I made some more experiments, to reproduce the problem, just try to os.fork lots of child processes, and import modules in the child process, and import fails with:

ImportError: PYZ entry 'xxxxxxx' failed to decompress

Here is an example:

``` python
import os
import time

def child():
  from pprint import pprint
  import hashlib
  from xml import etree
  pprint('Hello World')
  time.sleep(5)


def main():
  pids = []
  for _ in xrange(512):
    pid = os.fork()
    if pid != 0:
      # parent
      print 'New process:', pid
      pids.append(pid)
    else:
      child()
      return
  time.sleep(30)


if __name__ == '__main__':
  main()

```
Pyinstaller and run it, it will produce lot's of error message like:

```
  File "<string>", line 5, in child
  File "/home/hyt/work/pyinstaller/env/lib/python2.7/site-packages/PyInstaller/loader/pyi_importers.py", line 202, in load_module
    is_pkg, bytecode = self._pyz_archive.extract(fullname)
  File "/home/hyt/work/pyinstaller/env/lib/python2.7/site-packages/PyInstaller/loader/pyi_archive.py", line 315, in extract
    raise ImportError("PYZ entry '%s' failed to decompress" % name)
ImportError: PYZ entry 'pprint' failed to decompress

```

**Comment by htgoebel** [comment]
Thanks for this script. I was able to reproduce it even using the current developer head.

**Comment by htgoebel** [comment]
I think, I found the reason: `imp.acquire_lock()` acquires the interpreter's import lock for the current *thread*. But we have multi-processes here.

**Comment by romanows** [comment]
Thanks for this discussion.  It seems to me that the `Archive`/`ZlibArchive` classes hold an open file object that is shared by forked child processes.  If two processes are both calling `seek()` and `read()` at the same time, this can cause the data that is read to be corrupted.  When passed to `zlib.decompress()`, the corrupted data causes the import error to be raised.

Could we refactor the `ZlibArchive.extract()` method to open its own file object on each call?  This seems to solve the problem for me on Linux.  In the very brittle workaround below, I patch the `extract()` function such that each call first closes the `ZlibArchive` object's open file handle and then opens another.  It suffices to put the code somewhere in the parent thread before the child processes run.

```python
    frozen_importers = [x for x in sys.meta_path if "FrozenImporter" in str(type(x))]
    for frozen_importer in frozen_importers:
        old_extract = frozen_importer._pyz_archive.__class__.extract
        def new_extract(self, name):
            self.lib.close()                    # THIS IS THE WORKAROUND 1/2
            self.lib = open(self.path, 'rb')    # THIS IS THE WORKAROUND 2/2
            return old_extract(self, name)
        frozen_importer._pyz_archive.__class__.extract = new_extract
```

Just because I had a hard time finding this issue initially: my first view of the problem was intermittently getting the exception `SAXReaderNotAvailable: No parsers found` when trying to read multiple XML files in parallel using the Python multiprocessing module.

**Comment by htgoebel** [comment]
Note: I tested #1145 with the result of  https://github.com/pyinstaller/pyinstaller/issues/1070#issuecomment-77863852 above.

**Comment by Bluehorn** [comment]
I just tested the example above with commit 5967ae25ff5541d468eb36a0b640fc6dcdb6a77f and it is still broken. I don't grok why this ticket is closed?

**Comment by matysek** [comment]
@Bluehorn try the `python3` branch and report back.

**Comment by Bluehorn** [comment]
@matysek: I tried before writing but only got so far:

```
(py3)->torsten@horatio:~/LOCO-2097$ python3 -V
Python 3.5.0a4+
(py3)->torsten@horatio:~/LOCO-2097$ python3 ~/opensrc/pyinstaller3/pyinstaller.py fork_prob.py 
fatal: Not a git repository (or any parent up to mount point /home)
Stopping at filesystem boundary (GIT_DISCOVERY_ACROSS_FILESYSTEM not set).
Traceback (most recent call last):
  File "/home/torsten/opensrc/pyinstaller3/pyinstaller.py", line 17, in <module>
    from PyInstaller.main import run
  File "/home/torsten/opensrc/pyinstaller3/PyInstaller/main.py", line 20, in <module>
    import PyInstaller.building.build_main
  File "/home/torsten/opensrc/pyinstaller3/PyInstaller/building/build_main.py", line 33, in <module>
    from .api import PYZ, EXE, DLL, COLLECT, MERGE
  File "/home/torsten/opensrc/pyinstaller3/PyInstaller/building/api.py", line 24, in <module>
    from PyInstaller.archive.writers import ZlibArchiveWriter, CArchiveWriter
  File "/home/torsten/opensrc/pyinstaller3/PyInstaller/archive/writers.py", line 50, in <module>
    PYMAGIC = _frozen_importlib.MAGIC_NUMBER
AttributeError: module 'importlib._bootstrap' has no attribute 'MAGIC_NUMBER'
```
I'll retry with a current CPython.

**Comment by matysek** [comment]
> File "/home/torsten/opensrc/pyinstaller3/PyInstaller/archive/writers.py", line 50, in <module>
>     PYMAGIC = _frozen_importlib.MAGIC_NUMBER
> AttributeError: module 'importlib._bootstrap' has no attribute 'MAGIC_NUMBER'

This is a known issue with python 3.5.

**Comment by htgoebel** [comment]
@bluehorn: Python 3.5 ist not yet supported, see https://github.com/pyinstaller/pyinstaller/wiki#development-version

**Comment by matysek** [comment]
@Bluehorn yesterday it was implemented experimental python 3.5 support you might try that but you need to recompile the bootloader.
